Refactor selection handling

tobous opened the pull request:
Migrates the selection handling from a text offset based handling to a line and in-line offset based handling. This is the first step in a bigger migration to completely switch the position handling to the new positioning system. The migration is continued in #872.

This migration to the new positioning scheme is done to make the position handling offset independent. This is a necessary step on the way towards internally only working on normalized content, which in turn is a step on the way towards providing IDE cross-compatibility.

This PR still contains some TODOs.
- in Core `SelectionCollector`: Will be dealt with after text edit activities were migrated to new positioning scheme
- in Saros/I `EditorManager`: Has to be decided what to do

Otherwise, this PR should be ready to review. As it is quite large, reviewing the commits separately will probably be the best strategy. The actual refactoring only takes place in the second to last commit `[REFACTOR] Replace old positioning scheme for selection handling`. The commits before that are setup introducing new logic and classes used by the new selection handling.

### Commits

<details><summary><b>[REFACTOR] Rename TextSelection to OldTextSelection</b></summary>
<br>

Renames TextSelection to OldTextSelection in preparation of the upcoming
refactoring introducing a new TextSelection implementation that works
based on line numbers and in-line offsets instead of text offsets.

The current implementation, now named OldTextSelection, will be removed
as part of the refactoring.

Subsequently rename EditorAPI.getSelection(IEditorPart) to
getOldSelection(...). This was done to avoid clashes with the new
implementation using the same signature.

</details>

<details><summary><b>[INTERNAL][I] Introduce BackgroundEditorPool</b></summary>
<br>

Adds BackgroundEditorPool to provide access to editor objects for
resources that are not currently open in an editor.

This will be necessary for the upcoming refactoring moving to a line
number and in-line offset based system to described positions in a text
document.  As the local logic still operates on text offsets, we have to
convert between the new positions and text offsets. Editors offer an
easy API to do such conversions. Furthermore, the API works on a caching
model, which should speed up the operations substantially.

The pool is backed by an LRU cache to avoid holding to many editor
references that are potentially no longer necessary, thereby creating a
quasi memory leak during a session. Instead, only the last 20 requested
editors are held and swapped out as necessary.

Adds the necessary methods to instantiate and release such editors to
the ProjectAPI.

Integrates the BackgroundEditorPool with the rest of the Saros logic to
clean it up according to state changes:
 - drop background editor when a real one is opened by the user
 - drop background editor when resource is deleted
 - drop background editor when session ends
 - drop background editor before application is torn down (necessary to
   avoid IntelliJ memory leak detection)

Adjusts the IntelliJ context test setup to mock the necessary IntelliJ
components used by the BackgroundEditorPool.

</details>

<details><summary><b>[INTERNAL][I] Add method to only drop selection annotations</b></summary>
<br>

Adds a method to the AnnotationManager to just drop the current
selection annotation without adding a new one. This method will be
necessary for an upcoming refactoring and the general IDE
cross-compatibility as Eclipse can create empty selection annotations
that symbolize that there is no selection or cursor in the given editor.

</details>

<details><summary><b>[INTERNAL][CORE] Introduce new text position class</b></summary>
<br>

Introduces TextPosition which can be used to specify a text position
through the use of a line number and an in-line offset.

The class will be used to specify text positions in the upcoming
refactoring moving to the new positioning scheme.

</details>

<details><summary><b>[INTERNAL][CORE] Introduce new text selection class</b></summary>
<br>

Introduces TextSelection which can be used to specify a text selection
using two TextPosition objects.

This class will be used to specify text selections in the upcoming
refactoring moving to the new positioning scheme.

</details>

<details><summary><b>[INTERNAL] Add API calls to convert positioning schemes</b></summary>
<br>

Adds the necessary calls to the Saros/E and Saros/I EditorAPI to convert
between the text offset based and the line number and in-line offset
based positioning scheme.

These methods will be used in the upcoming refactoring migrating to the
line number and in-line offset based positioning scheme. The conversions
are vital for the refactoring as the rest of the Saros logic still works
on text offset based position data.

The logic introduced it backed by methods of the corresponding IDEs.
Hopefully, these methods are optimized internally to allow for a better
runtime.

</details>

<details><summary><b>[REFACTOR] Replace old positioning scheme for selection handling</b></summary>
<br>

Refactors the current logic dealing with selections to work with the new
line number and in-line offset based positioning scheme.

Replaces the usage of OldTextSelection with TextSelection.

Adjusts TextSelectionActivity to work with the new positioning scheme.
Subsequently improves the ease of use of the class by offering methods
that directly work on TextSelection objects.

Adjusts the tests dealing with selections. Chooses new values for test
selections. The used positions were chosen "randomly" (i.e. whatever
numbers came to mind).

Adjusts the EditorState to also work with the new positioning scheme.

Uses the EditorAPI methods to convert between the old, text offset based
positioning scheme and the new, line number and in-line offset based
positioning scheme. This approach was chosen to change as little logic
as possible during this refactoring. By converting, the rest of the
Saros logic dealing with selections internally can still be reused.
If this becomes an issue in the future, additional components currently
relying on the conversion can still be refactored to completely work
with the new positioning scheme in a future PR.

Comments out some logic in the FollowModeManager and SelectionCollector
where the old and new positioning schemes clash. These parts of the
logic take the offsets of a text edit activity and want to create a
selection from these offsets. As both classes are located in the core, a
conversion between the positioning schemes is not possible (without
doing it manually).
These parts of the logic will be re-added and adjusted once the text
edit activity handling is also moved to the new positioning scheme, at
which point the clash will be resolved as both parts will work with the
new positioning scheme.

</details>

<details><summary><b>[NOP] Remove classes and methods no longer needed after refactoring</b></summary>
<br>

Removes OldTextSelection as it has completely been replace by
TextSelection.

Removes no longer used methods in the EditorAPI classes that were only
used for the old selection handling.

</details>

<details><summary><b>[JUNIT] Introduce unit tests for the new position calculation logic</b></summary>
<br>

Adds unit tests for the new text position calculation logic introduced
to the Eclipse and IntelliJ implementation.

</details>

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 2
           

Complexity increasing per file
==============================
- core/src/saros/editor/text/TextPosition.java  4
- core/src/saros/editor/text/TextSelection.java  3
- eclipse/src/saros/editor/internal/EditorAPI.java  3
- intellij/src/saros/intellij/editor/EditorAPI.java  5
- intellij/src/saros/intellij/editor/BackgroundEditorPool.java  2
         

Complexity decreasing per file
==============================
+ eclipse/src/saros/feedback/SelectionCollector.java  -2
+ core/src/saros/activities/TextSelectionActivity.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5112097&bid=16808949)

tobous commented:
The STF also still passes with this Patch. See [build](https://github.com/saros-project/saros/runs/482657006).

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 2
           

Complexity increasing per file
==============================
- core/src/saros/editor/text/TextSelection.java  3
- intellij/src/saros/intellij/editor/EditorAPI.java  5
- intellij/src/saros/intellij/editor/BackgroundEditorPool.java  2
- eclipse/src/saros/editor/internal/EditorAPI.java  3
- core/src/saros/editor/text/TextPosition.java  4
         

Complexity decreasing per file
==============================
+ eclipse/src/saros/feedback/SelectionCollector.java  -2
+ core/src/saros/activities/TextSelectionActivity.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5112097&bid=16808949)

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 2
           

Complexity increasing per file
==============================
- intellij/src/saros/intellij/editor/EditorAPI.java  5
- core/src/saros/editor/text/TextSelection.java  3
- intellij/src/saros/intellij/editor/BackgroundEditorPool.java  2
- core/src/saros/editor/text/TextPosition.java  4
- eclipse/src/saros/editor/internal/EditorAPI.java  3
         

Complexity decreasing per file
==============================
+ core/src/saros/activities/TextSelectionActivity.java  -2
+ eclipse/src/saros/feedback/SelectionCollector.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5112097&bid=16808949)

m273d15 commented on eclipse/src/saros/editor/internal/EditorAPI.java:
```diff
@@ -291,6 +296,242 @@ public static OldTextSelection getOldSelection(IEditorPart editorPart) {
+
+    IDocument document = textEditor.getDocumentProvider().getDocument(editorPart.getEditorInput());
+
+    ITextSelection textSelection = (ITextSelection) selectionProvider.getSelection();
```
The [ITextSelection](https://help.eclipse.org/2019-12/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fapi%2Forg%2Feclipse%2Fjface%2Ftext%2FITextSelection.html) provides methods as `getStartLine` is this another context?

tobous commented on core/test/junit/saros/activities/ActivityOptimizerTest.java:
```diff
@@ -49,21 +51,24 @@ public void testOptimize() {
+    TextSelectionActivity tsChange0 = new TextSelectionActivity(alice, selection1, foofooSPath);
 
-    TextSelectionActivity tsChange2 = new TextSelectionActivity(alice, 0, 1, foobarSPath);
+    TextSelectionActivity tsChange1 = new TextSelectionActivity(alice, selection2, foofooSPath);
```
Actually, not a copy paste error but just a weird display in the diff. The diff is off by 1, leading to a situation where i changed the second parameter. If you compare the actual order of parameters side by side, you will find that they are still the same:
foofoo -> foofoo -> foobar -> foobar -> barfoo -> barfoo -> barbar -> barbar

tobous commented on eclipse/src/saros/editor/internal/EditorAPI.java:
```diff
@@ -291,6 +296,242 @@ public static OldTextSelection getOldSelection(IEditorPart editorPart) {
+      return TextSelection.EMPTY_SELECTION;
+    }
+
+    IDocument document = textEditor.getDocumentProvider().getDocument(editorPart.getEditorInput());
```
Done.

tobous commented on core/src/saros/activities/TextSelectionActivity.java:
```diff
@@ -21,48 +21,62 @@
-  protected final int length;
-
-  public TextSelectionActivity(User source, int offset, int length, SPath path) {
+  @XStreamAsAttribute private final int startLine;
```
Done. Used long, descriptive names.

m273d15 commented on intellij/src/saros/intellij/editor/BackgroundEditorPool.java:
```diff
@@ -0,0 +1,219 @@
+     * @param loadFactor the load factor to use
+     * @see LinkedHashMap#LinkedHashMap(int, float, boolean)
+     */
+    LRUCache(int capacity, float loadFactor) {
```
Aah, I was confused by the combination of both data structures, but the LinkedList is just an organizational structure. You are right 😄 However, I would store the constant LOAD_FACTOR in the LRU cache.

tobous replied:
Done.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 2
           

Complexity increasing per file
==============================
- core/src/saros/editor/text/TextSelection.java  3
- intellij/src/saros/intellij/editor/BackgroundEditorPool.java  2
- intellij/src/saros/intellij/editor/EditorAPI.java  5
- core/src/saros/editor/text/TextPosition.java  4
- eclipse/src/saros/editor/internal/EditorAPI.java  3
         

Complexity decreasing per file
==============================
+ core/src/saros/activities/TextSelectionActivity.java  -2
+ eclipse/src/saros/feedback/SelectionCollector.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5112097&bid=16808949)

m273d15 commented on intellij/src/saros/intellij/editor/EditorAPI.java:
```diff
@@ -103,6 +106,134 @@ public static LineRange getLocalViewPortRange(
+    int endLineOffset;
+
+    // Skip line offset calculation if end is in same line as start
+    if (endLine == startLine) {
```
Do as you like. It is just a matter of taste.

tobous replied:
I think I like it better the way it is to keep the handling the same for all cases.

tobous commented:
@m273d15 Bad timing with the approval. :wink: Still had to push the LRU load factor adjustment. Also, the STF not re-runs to make sure.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 2
           

Complexity increasing per file
==============================
- core/src/saros/editor/text/TextPosition.java  4
- intellij/src/saros/intellij/editor/EditorAPI.java  5
- intellij/src/saros/intellij/editor/BackgroundEditorPool.java  2
- eclipse/src/saros/editor/internal/EditorAPI.java  3
- core/src/saros/editor/text/TextSelection.java  3
         

Complexity decreasing per file
==============================
+ eclipse/src/saros/feedback/SelectionCollector.java  -2
+ core/src/saros/activities/TextSelectionActivity.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5112097&bid=16808949)

tobous commented on core/src/saros/activities/TextSelectionActivity.java:
```diff
@@ -21,48 +21,72 @@
 public class TextSelectionActivity extends AbstractResourceActivity {
 
-  @XStreamAlias("o")
+  @XStreamAlias("selectionStartLine")
```
Incompatible with what? With previous releases? Isn't that already the case with me just adding new parameters?

If you would like to minimize packet size, I can shorten them to two chars. Would that be acceptable?

srossbach replied:
Yes, just keep it short, the alias is only for XStream.

tobous commented on core/src/saros/activities/TextSelectionActivity.java:
```diff
@@ -21,48 +21,72 @@
+   * @param path the resource the text was selected in
+   * @throws IllegalArgumentException if the source or path is <code>null</code>
+   */
+  public TextSelectionActivity(User source, TextSelection selection, SPath path) {
```
The reason is that, while a selection can be seen as an interval of chars, the offset and length of this interval is dependent on the local line endings. This is a problem when trying to work with other participants using different line endings.

To circumvent this issue, there are two viable options as far as I can tell:
1. Normalize offsets locally (i.e. calculate what the corrected offsets and length would be if the file were using Unix line ending) and then revert the normalization (if needed) on the receiving side.
2. Calculate the line and in-line offset of both the start and end of the selection (as this is a line-ending independent way of specifying a position in the document) and re-convert it on the receiving side.

(Theoretically we could require all shared documents to use Unix line endings, but that would exclude many potential use cases and severely limit Saros' usability.)

Both approaches share the same base idea to use a unified format.

The advantage of approach 1 is that we don't create any overhead in cases where we or the receiving side already use the format chosen as the internal standard (e.g. Unix line endings).
The disadvantage is that we have to make the conversion ourselves by going through the whole document up to the needed offsets and counting line endings.

The advantage of approach 2 is that the IDE implementations already offer (optimized) methods to convert between the offset and line format.
This disadvantage is that the conversion always has to be done, even if the format of all participants already matches (meaning no normalization would be necessary).

Weighing these options, I decided to choose the option 2 as it seemed the most sensible to me.

srossbach commented on eclipse/src/saros/editor/internal/EditorAPI.java:
```diff
@@ -268,27 +272,275 @@ public static IResource getEditorResource(IEditorPart editorPart) {
+      return org.eclipse.jface.text.TextSelection.emptySelection();
+    }
+
+    int startOffset = startLineOffset + startPosition.getInLineOffset();
```
You should add a sanity check because there is now way compare TextLocations.

Eg. You can have a TextLocation(0, 20) and TextLocation(1, 5). The first TextLocation is invalid because the first line was altered in some way only only contains 4 characters now. In the end you will end up with a negative length. This can also happen if you just transmit faulty values.

tobous replied:
Yes, I agree that a sanity check would be nice when calculating the selection values. I will think about how best to do this.

But I also don't think that this is an issue new to then new handling. As selections are not processed as part of the jupiter operation stuff, we could always run into situations where a local selection can no longer be applied correctly as the document has changed.

srossbach replied:
My guess is that the annotation position and or offset does not really have to match the current document boundaries the annotation is put in. As long as no exception is triggered I am fine with some log statements (warning or debug). You already mentioned that normally selection must also use the Jupiter algorithm but we do not do that.

tobous replied:
An additional sanity check here should not be necessary. If the given offsets/text positions can no longer be converted, the logic throws a `BadLocationException`. This exception is already caught internally, logged and an empty selection is returned instead.

And, in the unlikely situation that the line offset is valid but the in-line offset pushes the calculated offset out of bounds, the annotation will be dropped silently when trying to apply it to the local model. See [Eclipse documentation](https://help.eclipse.org/2018-09/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fapi%2Forg%2Feclipse%2Fjface%2Ftext%2Fsource%2Fclass-use%2FIAnnotationModelExtension.html) (used in our code [here](https://github.com/saros-project/saros/blob/master/eclipse/src/saros/editor/internal/AnnotationModelHelper.java#L65)).

For IntelliJ, such cases are [also handled by the annotation](https://github.com/saros-project/saros/blob/master/intellij/src/saros/intellij/editor/annotations/AnnotationManager.java#L713) logic.

srossbach commented:
Can this PR be delayed until the Jupiter overhaul is done or must it be merged right now ? I know it do no harm but it might looks confusing if there are issues with the Jupiter stuff and so the overhaul could be delayed for a good amount of time.

tobous commented:
> Can this PR be delayed until the Jupiter overhaul is done or must it be merged right now ? I know it do no harm but it might looks confusing if there are issues with the Jupiter stuff and so the overhaul could be delayed for a good amount of time.

I wanted to wait until I was done with the base prototype of the jupiter refactoring anyways. So yes, I am not in a hurry to merge this until it is certain that the approach completely works.

srossbach commented on eclipse/src/saros/editor/EditorListener.java:
```diff
@@ -201,8 +201,13 @@ protected void generateSelection() {
 
-      manager.generateSelection(
-          part, new saros.editor.text.TextSelection(selection.getOffset(), selection.getLength()));
+      IDocument document = viewer.getDocument();
```
Can you please add an assertion that the document is never null ? Reading the Interface Javadoc it could. I do not know if this still holds true for textviewers.

tobous replied:
Will add a check logging a warning and using `EMPTY_SELECTION` instead if the document is `null`.

srossbach commented on eclipse/src/saros/editor/internal/EditorAPI.java:
```diff
@@ -268,27 +272,275 @@ public static IResource getEditorResource(IEditorPart editorPart) {
-   * @param editorPart the editorPart for which to get the text selection.
-   * @return the current text selection. Returns {@link TextSelection#emptySelection()} if no text
-   *     selection exists.
+   * <p>The given editor part must not be <code>null</code>.
```
There was a hidden null check that returned an empty selection instead of throwing an exception.

tobous replied:
Yeah, it is still there l.290. Will remove the `Objects.requireNonNull(...)` and update the javadoc.

srossbach commented on eclipse/src/saros/editor/internal/EditorAPI.java:
```diff
@@ -268,27 +272,275 @@ public static IResource getEditorResource(IEditorPart editorPart) {
+   *     not be calculated
+   * @throws NullPointerException if the given editor part is <code>null</code>
    */
   public static TextSelection getSelection(IEditorPart editorPart) {
```
I know that this is not your fault but the EditorAPI is basically just convenience API regarding Eclipse stuff.

Normally it would be better to return a JFace TextSelection, to be more specific an ITextSelection and then introduce fromCoreToEclipse and fromEclipseToCore methods to convert them.

tobous replied:
But why? All these methods do is use Eclipse specific API to calculate some values. And yes, the input and output values are wrapped in a class from the "core", but why does it matter? I could also pass them as an array of `int`, but that just seem cumbersome to me.

Under the premise that we are working with the new positioning logic, I don't see a need to access the old selection format. So why keep it in a separate method?

srossbach replied:
As I said it is generic API. Every method beside ```getEditorPath``` is taking and returning Eclipse stuff.

tobous replied:
So what would you prefer me doing? Create a new utility class named `SelectionAPI` (or something like that) and move the methods there?

srossbach replied:
Just add an TODO. The refactor / movement of the logic is not urgent.

tobous replied:
Done.

srossbach commented on eclipse/src/saros/editor/internal/EditorAPI.java:
```diff
@@ -268,27 +272,275 @@ public static IResource getEditorResource(IEditorPart editorPart) {
+   * @throws NullPointerException if the given editor part or selection is <code>null</code>
+   * @throws IllegalArgumentException if the given text selection is empty
+   */
+  public static ITextSelection calculateOffsets(IEditorPart editorPart, TextSelection selection) {
```
I guess both methods can easily be JUnit tested ?

tobous replied:
No, not really...We would need a working eclipse backing or would have to mock the necessary methods (like the requests to `IDocument`, `SelectionProvider`, etc).

srossbach replied:
I do not think that ``` public static TextSelection calculateSelection(IDocument document, int offset, int length)``` is hard to test. Even this method is not that hard to test. You need to setup 3 mocks where 2 mocks do not really do anything.

tobous replied:
I added tests for both methods for both the Eclipse and IntelliJ implementation.

It is way overbuild in my opinion, but it is a test suite.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 2
           

Complexity increasing per file
==============================
- eclipse/src/saros/editor/internal/EditorAPI.java  3
- intellij/src/saros/intellij/editor/BackgroundEditorPool.java  2
- core/src/saros/editor/text/TextSelection.java  3
- intellij/src/saros/intellij/editor/EditorAPI.java  5
- core/src/saros/editor/text/TextPosition.java  4
         

Complexity decreasing per file
==============================
+ core/src/saros/activities/TextSelectionActivity.java  -2
+ eclipse/src/saros/feedback/SelectionCollector.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5112097&bid=16808949)

tobous commented:
Rebased onto the current master.

Addressed review comments:
- added TODO to move methods in Eclipse `EditorAPI` to separate class 
- added test suite for new `EditorAPI` methods

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 7
           

Complexity increasing per file
==============================
- intellij/src/saros/intellij/editor/EditorAPI.java  5
- eclipse/src/saros/editor/internal/EditorAPI.java  3
- core/src/saros/editor/text/TextPosition.java  4
- core/src/saros/editor/text/TextSelection.java  3
- eclipse/test/junit/saros/editor/internal/EditorAPITextPositionTest.java  3
- intellij/test/junit/saros/intellij/editor/EditorAPITextPositionTest.java  3
- intellij/src/saros/intellij/editor/BackgroundEditorPool.java  2
         

Complexity decreasing per file
==============================
+ eclipse/src/saros/feedback/SelectionCollector.java  -2
+ core/src/saros/activities/TextSelectionActivity.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5112097&bid=16808949)

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 3
           

Complexity increasing per file
==============================
- intellij/src/saros/intellij/editor/BackgroundEditorPool.java  2
- eclipse/src/saros/editor/internal/EditorAPI.java  3
- intellij/test/junit/saros/intellij/editor/EditorAPITextPositionTest.java  3
- core/src/saros/editor/text/TextPosition.java  4
- core/src/saros/editor/text/TextSelection.java  3
- eclipse/test/junit/saros/editor/internal/EditorAPITextPositionTest.java  3
- intellij/src/saros/intellij/editor/EditorAPI.java  5
         

Complexity decreasing per file
==============================
+ eclipse/src/saros/feedback/SelectionCollector.java  -2
+ core/src/saros/activities/TextSelectionActivity.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5112097&bid=16808949)

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 3
           

Complexity increasing per file
==============================
- core/src/saros/editor/text/TextPosition.java  4
- intellij/test/junit/saros/intellij/editor/EditorAPITextPositionTest.java  3
- intellij/src/saros/intellij/editor/BackgroundEditorPool.java  2
- eclipse/test/junit/saros/editor/internal/EditorAPITextPositionTest.java  3
- core/src/saros/editor/text/TextSelection.java  3
- intellij/src/saros/intellij/editor/EditorAPI.java  5
- eclipse/src/saros/editor/internal/EditorAPI.java  3
         

Complexity decreasing per file
==============================
+ core/src/saros/activities/TextSelectionActivity.java  -2
+ eclipse/src/saros/feedback/SelectionCollector.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5112097&bid=16808949)

tobous commented:
@m273d15 @srossbach Are the test suites satisfactory? Do you have any other issues with this PR? (As a reminder, I won't merge this until the second part of the refactoring is approved and ready to be merged.)

m273d15 commented:
@tobous You asked for a review of your tests. I wanted to describe my approach by text, but recognized that it would be to wordy. Therefore, I created a corresponding [commit](https://github.com/saros-project/saros/compare/pr/core/refactor-selection-offset-handling...pr/core/rsoh_test?expand=1) that contains the main idea.

My main idea to improve the maintainability:
* If you are not testing the validity check of the "calculateOffsets" method then skip the check with separating the logic.
* Using a builder to create the mock instead of using the complex structure.
* Showing relations between the numbers by using variables. I think I missed a relation in the selection calculation.

tobous commented:
I just noticed that i forgot to add the assertion messages. Will either add some or remove the empty messages.

> @tobous You asked for a review of your tests. I wanted to describe my approach by text, but recognized that it would be to wordy. Therefore, I created a corresponding [commit](https://github.com/saros-project/saros/compare/pr/core/refactor-selection-offset-handling...pr/core/rsoh_test?expand=1) that contains the main idea.

In general, it seems like a valid approach, but I am not sure whether the larger restructuring really improves much. (See comments below.)

> My main idea to improve the maintainability:
> 
> * If you are not testing the validity check of the "calculateOffsets" method then skip the check with separating the logic.

I am strongly against adding API methods (even if they are just `protected`) just to make the testing process easier. It just pollutes the API in my opinion.

But even with the `IDocumentBuilder` (or then `EditorPartBuilder`) approach, the additional mocking logic is not a big deal as it could still be hidden inside the builder.

> * Using a builder to create the mock instead of using the complex structure.

This approach works as well, but I don't see much of an improvement. The setup is still basically the same, except that you also always have to instantiate the document builder and document instance. 
And you still have to look at what the helper structure (either the builder or the maps and answer functions) to understand how to correctly use the setup (or just C&P and change values, which also works for both cases).

So I don't really care which approach is used, but I would question whether it is worth the time/effort to change the current setup for both test suites.

> * Showing relations between the numbers by using variables. I think I missed a relation in the selection calculation.

Seems like a good idea. I will definitely adopt it.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 3
           

Complexity increasing per file
==============================
- intellij/test/junit/saros/intellij/editor/EditorAPITextPositionTest.java  3
- eclipse/test/junit/saros/editor/internal/EditorAPITextPositionTest.java  3
- core/src/saros/editor/text/TextSelection.java  3
- intellij/src/saros/intellij/editor/EditorAPI.java  5
- core/src/saros/editor/text/TextPosition.java  4
- intellij/src/saros/intellij/editor/BackgroundEditorPool.java  2
- eclipse/src/saros/editor/internal/EditorAPI.java  3
         

Complexity decreasing per file
==============================
+ eclipse/src/saros/feedback/SelectionCollector.java  -2
+ core/src/saros/activities/TextSelectionActivity.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5112097&bid=16808949)

tobous commented:
Introduced variables for values used multiple times in test cases. Slightly restructures tests to make them easier to read.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 3
           

Complexity increasing per file
==============================
- core/src/saros/editor/text/TextPosition.java  4
- intellij/test/junit/saros/intellij/editor/EditorAPITextPositionTest.java  3
- intellij/src/saros/intellij/editor/EditorAPI.java  5
- intellij/src/saros/intellij/editor/BackgroundEditorPool.java  2
- eclipse/test/junit/saros/editor/internal/EditorAPITextPositionTest.java  3
- eclipse/src/saros/editor/internal/EditorAPI.java  3
- core/src/saros/editor/text/TextSelection.java  3
         

Complexity decreasing per file
==============================
+ core/src/saros/activities/TextSelectionActivity.java  -2
+ eclipse/src/saros/feedback/SelectionCollector.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5112097&bid=16808949)

m273d15 commented:
> I am strongly against adding API methods (even if they are just protected) just to make the testing process easier.

I would agree with your statement if we would talk about introducing extra logic for tests into the production code, but we are talking about code structure.

Splitting a large method that is hard to read into multiple smaller methods is always a good idea.
In order to make the code more maintainable you want to reduce the complexity (which means the time and effort to understand the code and change it).
In a refactoring I would try to split the whole method into multiple methods (e.g. one method for each concern: verifying the input parameters, determine document, calculate the offsets). I don't follow this dogmatically, but I think it is a good approach.

Introducing a new method makes boundaries and data flow more clear, e.g. in this context, splitting the method into two methods would allow the reader to directly understand that the actual calculation only needs the IDocument and the IEditorPart is only required for getting the IDocument and is not needed for understanding the calculation (I can remove it from my mental list of references).

> It just pollutes the API in my opinion.

I would not define the API of a class as the set of all class methods but as the set of methods that are exposed to the API user (in this case the public methods).
If the API definition mechanism does not allow you to add methods for internal restructuring, then (my opinion) the definition (or the mechanism) is broken and we would need to adjust it.

> But even with the IDocumentBuilder (or then EditorPartBuilder) approach, the additional mocking logic is not a big deal as it could still be hidden inside the builder.

I know you already wrote the mocks, but if it is possible to avoid additional mocks in a test I would do it. These mocks confuse the reader. I had to understand that they
are not required for triggering an effect you want to test. The mocks are just for satisfying the first part of your method (the verify and determine document part). And if you want to test your first part, you could create a separate method for it and create all the mocks in one test instead of in all tests.

> This approach works as well, but I don't see much of an improvement. The setup is still basically the same, except that you also always have to instantiate the document builder and document instance.

Sure the setup is the same, but all I want to do is optimizing the readability. I see the following advantages concerning readability:
* The builder pattern is well known and as soon as you read "Builder" you know the general structure and that this class creates an IDocument.
* If I read a test using the builder I directly know that the values defined with the `withXY` method are related to the document (because the IDocument is created by the builder and the `withXY` method is called on the builder),
in the current version you first have to understand that the values that are added to the maps will be returned by the document. Furthermore, you
first have to understand how the IAnswer callback works.

> So I don't really care which approach is used, but I would question whether it is worth the time/effort to change the current setup for both test suites.

As you like, do you think this would take a lot of time?

tobous commented:
> > I am strongly against adding API methods (even if they are just protected) just to make the testing process easier.
> 
> I would agree with your statement if we would talk about introducing extra logic for tests into the production code, but we are talking about code structure.
> 
> Splitting a large method that is hard to read into multiple smaller methods is always a good idea.
> In order to make the code more maintainable you want to reduce the complexity (which means the time and effort to understand the code and change it).
> In a refactoring I would try to split the whole method into multiple methods (e.g. one method for each concern: verifying the input parameters, determine document, calculate the offsets). I don't follow this dogmatically, but I think it is a good approach.
> 
> Introducing a new method makes boundaries and data flow more clear, e.g. in this context, splitting the method into two methods would allow the reader to directly understand that the actual calculation only needs the IDocument and the IEditorPart is only required for getting the IDocument and is not needed for understanding the calculation (I can remove it from my mental list of references).

I completely agree with the notion that it is useful to split logic into different methods, but I would generally try to keep such a split internal (i.e. using the private modifier) if the created "partial" methods are not used otherwise.

> > It just pollutes the API in my opinion.
> 
> I would not define the API of a class as the set of all class methods but as the set of methods that are exposed to the API user (in this case the public methods).
> If the API definition mechanism does not allow you to add methods for internal restructuring, then (my opinion) the definition (or the mechanism) is broken and we would need to adjust it.

For me, the API would be everything that can be accessed from outside the class. And that also includes protected methods. Having to many visible methods just makes it hard to get an overview of what a class actually does and to find the methods you are actually looking for.

But maybe that viewpoint is too strict.

I didn't notice that the class is in the `internal` package, which makes the usage of package protected scope more acceptable. But still, I don't think it is worth it spiting the method in this particular case, especially because the skipped logic is just boilerplate code.

> > But even with the IDocumentBuilder (or then EditorPartBuilder) approach, the additional mocking logic is not a big deal as it could still be hidden inside the builder.
> 
> I know you already wrote the mocks, but if it is possible to avoid additional mocks in a test I would do it. These mocks confuse the reader. I had to understand that they
> are not required for triggering an effect you want to test. The mocks are just for satisfying the first part of your method (the verify and determine document part). And if you want to test your first part, you could create a separate method for it and create all the mocks in one test instead of in all tests.
> 
> > This approach works as well, but I don't see much of an improvement. The setup is still basically the same, except that you also always have to instantiate the document builder and document instance.
> 
> Sure the setup is the same, but all I want to do is optimizing the readability. I see the following advantages concerning readability:
> 
> * The builder pattern is well known and as soon as you read "Builder" you know the general structure and that this class creates an IDocument.
> * If I read a test using the builder I directly know that the values defined with the `withXY` method are related to the document (because the IDocument is created by the builder and the `withXY` method is called on the builder),
>   in the current version you first have to understand that the values that are added to the maps will be returned by the document. Furthermore, you
>   first have to understand how the IAnswer callback works.
> 
> > So I don't really care which approach is used, but I would question whether it is worth the time/effort to change the current setup for both test suites.
> 
> As you like, do you think this would take a lot of time?

Replaced previous logic with builder setup.

tobous commented:
Rebased onto current master.

tobous commented:
@srossbach Do you still have any issues with this PR? Otherwise I would see this PR as completely reviewed. (Again, I am still not going to merge it until the rest is reviewed.)

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 4
           

Complexity increasing per file
==============================
- core/src/saros/editor/text/TextSelection.java  3
- intellij/test/junit/saros/intellij/editor/EditorAPITextPositionTest.java  1
- intellij/src/saros/intellij/editor/EditorAPI.java  5
- eclipse/test/junit/saros/editor/internal/EditorAPITextPositionTest.java  2
- core/src/saros/editor/text/TextPosition.java  4
- eclipse/src/saros/editor/internal/EditorAPI.java  3
- intellij/src/saros/intellij/editor/BackgroundEditorPool.java  2
         

Complexity decreasing per file
==============================
+ core/src/saros/activities/TextSelectionActivity.java  -2
+ eclipse/src/saros/feedback/SelectionCollector.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5112097&bid=16808949)

tobous commented:
Rebased onto current master, resolving some minor merge issues caused by the logger refactoring.

stefaus commented on intellij/src/saros/intellij/editor/ProjectAPI.java:
```diff
@@ -16,6 +17,12 @@
+
+  static {
+    editorFactory = EditorFactory.getInstance();
+  }
```
why this split?

stefaus commented on intellij/src/saros/intellij/editor/LocalEditorManipulator.java:
```diff
@@ -243,13 +246,8 @@ public void adjustViewport(@NotNull Editor editor, LineRange range, TextSelectio
-
-      remoteStartLine = selectionRange.getStartLine();
-      remoteEndLine = selectionRange.getStartLine() + selectionRange.getNumberOfLines();
+      remoteStartLine = selection.getStartPosition().getLineNumber();
```
`selection` may be null here

tobous replied:
But if the range is not `null`, the `if` case in line 244 is used, is it not?

stefaus replied:
Ok eclipse shows just a potential npe and github collapsed the diff in a excellent way to show it as possible. nice :D

stefaus commented on eclipse/test/junit/saros/editor/internal/EditorAPITextPositionTest.java:
```diff
@@ -0,0 +1,335 @@
+        EasyMock.expect(document.getLineOffset(lineNumberInput)).andReturn(lineOffsetAnswer);
+
+      } catch (BadLocationException ignored) {
+      }
```
please make the eclipse problem marker happy and add NOP comments :)

tobous replied:
This is handled with the name "ignored" for the IntelliJ analyze. But I am more than happy to also add a comment.

tobous commented on eclipse/src/saros/editor/internal/EditorAPI.java:
```diff
@@ -267,28 +271,274 @@ public static IResource getEditorResource(IEditorPart editorPart) {
+
+    if (documentProvider == null) {
+      return org.eclipse.jface.text.TextSelection.emptySelection();
+    }
```
I think that this is heavily subjective. I am not consistent on the handling either.

For me, this is a separate path of the code flow (the early exit for no valid document provider) and therefore is in its own block. But I have to admit that I sometimes overdo it with the blank lines in my code.

I still don't think that it is necessary to change this here, especially because this format matches the styling of (most of) the rest of the file.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 4
           

Complexity increasing per file
==============================
- eclipse/test/junit/saros/editor/internal/EditorAPITextPositionTest.java  2
- intellij/src/saros/intellij/editor/EditorAPI.java  5
- core/src/saros/editor/text/TextSelection.java  3
- core/src/saros/editor/text/TextPosition.java  4
- eclipse/src/saros/editor/internal/EditorAPI.java  3
- intellij/test/junit/saros/intellij/editor/EditorAPITextPositionTest.java  1
- intellij/src/saros/intellij/editor/BackgroundEditorPool.java  2
         

Complexity decreasing per file
==============================
+ eclipse/src/saros/feedback/SelectionCollector.java  -2
+ core/src/saros/activities/TextSelectionActivity.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5112097&bid=16808949)

tobous commented on intellij/src/saros/intellij/editor/BackgroundEditorPool.java:
```diff
@@ -0,0 +1,220 @@
+   * @see LinkedHashMap
+   * @see LinkedHashMap#removeEldestEntry(Map.Entry)
+   */
+  private static class LRUCache extends LinkedHashMap<SPath, Editor> {
```
Done.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 4
           

Complexity increasing per file
==============================
- intellij/src/saros/intellij/editor/BackgroundEditorPool.java  2
- intellij/test/junit/saros/intellij/editor/EditorAPITextPositionTest.java  1
- eclipse/test/junit/saros/editor/internal/EditorAPITextPositionTest.java  2
- core/src/saros/editor/text/TextSelection.java  3
- intellij/src/saros/intellij/editor/EditorAPI.java  5
- eclipse/src/saros/editor/internal/EditorAPI.java  3
- core/src/saros/editor/text/TextPosition.java  4
         

Complexity decreasing per file
==============================
+ core/src/saros/activities/TextSelectionActivity.java  -2
+ eclipse/src/saros/feedback/SelectionCollector.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5112097&bid=16808949)

tobous commented:
As all PRs in this chain dealing with the migration to using the new text positioning scheme and normalizing line endings for text edits (#862, #872, #874, #877, and #878) have been approved, I would like to proceed merging the chain in the near future.  (Technically, this PR still has a "changes requested" marker, but all requested changes have been addressed in my opinion. If anybody feels otherwise, please say so.)

(In the merging process, I am also going to do a final minimal adjustment for this PR by adding a private CTOR to the static utility class `TextPositionUtils`.)

My current plan is to merge this the upcoming Thursday, April 2nd.

@m273d15 @stefaus @srossbach If you have any objections or issues that you still would like to see addressed before merging, please say so beforehand.

tobous commented:
Rebased onto current master without any interaction.
